pagefaultgames/pokerogue
# [Feature] Implement Stockpile, Spit Up, Swallow

## What are the changes?
Implements Stockpile, Spit Up, and Swallow.

## Why am I doing these changes?
Because these moves are interesting.  

The other PRs (https://github.com/pagefaultgames/pokerogue/pull/2954 and https://github.com/pagefaultgames/pokerogue/pull/410) are either partial implementations (the former) or missing some key stat tracking logic.

## What did change?
These implementations rely on a new BattlerTag, named `StockpilingTag`.  The new tag:
- Handles stat changes on adding a stack, or removing all stacks.
- Keeps track of two pieces of state:
  1. The current number of Stockpile stacks
  2. The number of successful changes to DEF and SPDEF (tracked individually) from adding stacks.

In order to support this:
> ii. The number of successful changes to DEF and SPDEF (tracked individually) from adding stacks.

A new (optional) parameter was added to the constructor of `StatChangePhase` -- `onChange: StatChangeCallback`.

```ts
export type StatChangeCallback = (target: Pokemon, changed: BattleStat[], relativeChanges: number[]) => void
```

`changed` contains a list of the `BattleStat`s that will change.
`relativeChanges` is a mapping of `changed` to the relative change in level for each of its (`BattleStat`) entries.

For example, if DEF will increase by 2, and SPDEF will increase by 1, the callback will be invoked like:
```ts
this.onChange(this.getPokemon(), [BattleStat.DEF, BattleStat.SPDEF], [2, 1])
```

`StockpilingTag` provides a callback that increments its internal stat-tracking state for DEF and/or SPDEF if a non-zero value is present in `relativeChanges`.

### Screenshots/Videos

https://github.com/pagefaultgames/pokerogue/assets/54485715/9efc19a8-a06c-430d-899c-577e98f36318

(I got trolled by eggs hatching, no need to watch past the ~1:50 mark)

## How to test the changes?
Most of the simple scenarios are pretty easy to replicate by starting with Cramorant and playing like normal.  Make sure that:
- Spit Up with 1/2/3 stacks has a power of 100/200/300, respectively (use the console)
- Swallow with 1/2/3 heals by 25%/50%/100% (use your eyes)
- Spit Up and Swallow both fail if you have no Stockpile stacks
- Stockpile stacks (and tracked stat changes) carry over as expected between wild pokemon encounters

Overrides and/or additional log messages are likely the easiest way to test edge cases. For example:
- Simple/Contrary interactions
- Stockpile should work (increment stacks if possible) even if both DEF and SPDEF are at max
- Not counting (per stat) stat changes for any stacks that did not raise DEF and/or SPDEF due to max thresholds

I'm looking into the testing framework; if automated tests don't make it into this PR, I'll add them as a fast-follow.

## Checklist
- [x] There is no overlap with another PR? (mentioned above)
  - [x] Have I provided screenshots/videos of the changes? (open to adding more if need be)
